src/testServerController.ts

```diff
@@ -59,65 +58,43 @@ export class TestServerController implements vscodeTypes.Disposable {
           FORCE_COLOR: '1',
         };
       },
-      clientFactory: () => new TestServer(this._vscode),
       dumpIO: false,
+      onClose: () => {
+        this._connectionPromises.delete(config.configFile);
+      },
+      onError: error => {
+        this._connectionPromises.delete(config.configFile);
+      },
     });
-    const testServer = await testServerBackend.start();
+    if (!wsEndpoint)
+      return null;
+    const testServer = new TestServerConnection(wsEndpoint);
+    await testServer.connect();
+    this._connectionPromises.set(config.configFile, Promise.resolve(testServer));
```

Comment:
Do you want to set the promise at the start of this function, to avoid races when calling it twice before connection is done?

Reply:
Oh yes